llvm/lib/Analysis/LoopAccessAnalysis.cpp

```diff
@@ -412,12 +412,10 @@ bool RuntimePointerChecking::needsChecking(
 /// Return nullptr in case we couldn't find an answer.
 static const SCEV *getMinFromExprs(const SCEV *I, const SCEV *J,
                                    ScalarEvolution *SE) {
-  const SCEV *Diff = SE->getMinusSCEV(J, I);
```

Comment:
Off topic - but this is essentially simplifying a three way compare.  We've recently added three way compares to IR, and we've got analogous simplification logic inside SCEV for min/max.  Should we maybe have an interface on SCEV which does the 3 way compare, and returns either the result or a SCEVCouldNotCompute?

Reply:
Off topic, but in a slightly different way... It looks like the sole user of this is addPointer.  I can't see any reason that the Min and Max can't simply be the SCEV expressions for the min and max of all pointers.  We'd need to cost whether generating such a group is profitable, but that's distinct from legality.  One concern is that I think the current logic is relying on the min/max simplification to fail when two pointers are based on different underlying objects.  We'd need to explicitly handle that.  There may also be a profitability concern about forming "too large" a group if the offsets are runtime values with non-overlapping ranges.

Reply:
I think taking a normal SCEV min/max would be fine in terms of correctness, but as you say, profitability is less clear. I also suspect that the logic used here may be stronger than what SCEV does for this particular case. SCEV only uses isKnownViaNonRecursiveReasoning to simplify min/max expression. (Of course, it can also be weaker in other respects...)